src/imitation/algorithms/mce_irl.py

```diff
@@ -321,49 +321,92 @@ def __init__(
-
-            if isinstance(demonstrations, Iterable):
-                first_item = next(iter(demonstrations))
-                if isinstance(first_item, types.Trajectory):
-                    # Demonstrations are trajectories.
-                    for traj in demonstrations:
-                        # TODO(adam): vectorize?
-                        cum_discount = 1
-                        for obs in traj.obs:
-                            self.demo_state_om[obs.astype(bool)] += cum_discount
-                            cum_discount *= self.discount
-                else:
-                    # Demonstrations are a Torch DataLoader or other Mapping iterable
-                    for batch in demonstrations:
-                        self._set_demo_from_obs(batch["obs"])
-
-            elif isinstance(demonstrations, types.TransitionsMinimal):
-                self._set_demo_from_obs(demonstrations.obs)
-            else:
-                raise TypeError(
-                    f"Unsupported demonstration type {type(demonstrations)}",
+            return
+
+        # Demonstrations are either trajectories or transitions;
+        # we must compute occupancy measure from this.
+        self.demo_state_om = np.zeros((self.env.n_states,))
+
+        if isinstance(demonstrations, Iterable):
+            first_item = next(iter(demonstrations))
+            if isinstance(first_item, types.Trajectory):
+                self._set_demo_from_trajectories(demonstrations)
+                return
+
+        # Demonstrations are from some kind of transitions-like object. This does
+        # not contain timesteps, so can only compute OM when undiscounted.
+        if self.discount != 1.0:
+            raise ValueError(
+                "Cannot compute discounted OM from timeless Transitions.",
+            )
+
+        if isinstance(demonstrations, types.Transitions):
+            self._update_demo_from_obs(
+                demonstrations.obs,
+                demonstrations.dones,
+                demonstrations.next_obs,
+            )
+        elif isinstance(demonstrations, types.TransitionsMinimal):
+            self._update_demo_from_obs(demonstrations.obs, None, None)
+        elif isinstance(demonstrations, Iterable):
+            # Demonstrations are a Torch DataLoader or other Mapping iterable
+            for batch in demonstrations:
+                assert isinstance(batch, Mapping)
+                self._update_demo_from_obs(
+                    batch["obs"],
+                    batch.get("dones"),
+                    batch.get("next_obs"),
                 )
+        else:
+            raise TypeError(
+                f"Unsupported demonstration type {type(demonstrations)}",
+            )
 
-            self.demo_state_om /= self.demo_state_om.sum()  # normalize
+        # Normalize occupancy measure estimates
+        self.demo_state_om *= (self.env.horizon + 1) / self.demo_state_om.sum()
```

Comment:
IMO this belongs in `_update_demo_from_obs`; it's hard to see why this line is correct when the OM update logic is in another function. You could still accommodate the `isinstance(demonstrations, Iterable)` branch by chaining together the batches in `demonstrations` and making just one `_update_demo_from_obs` call.

Reply:
(also if you change this then you may as well change `_update_demo_from_obs` and `_update_demo_from_trajectories` to zero out `demo_state_om` at the beginning of each call)

Reply:
Changed. The collation code does make things a bit more complicated, but it's probably overall still easier to understand.